programs/whirlpool/src/util/sparse_swap.rs

```diff
@@ -0,0 +1,1746 @@
+
+    // following process is ported from anchor-lang's AccountLoader::try_from and AccountLoader::load_mut
+
+    if !account_info.is_writable {
+        return Err(anchor_lang::error::ErrorCode::AccountNotMutable.into());
+    }
+
+    // uninitialized writable account (owned by system program and its data size is zero)
+    if account_info.owner == &System::id() && account_info.data_is_empty() {
+        return Ok(TickArrayAccount::Uninitialized {
+            pubkey: *account_info.key,
+            account_info,
+            start_tick_index: None,
+        });
+    }
+
+    // owner program check
+    if account_info.owner != &TickArray::owner() {
+        return Err(
+            Error::from(anchor_lang::error::ErrorCode::AccountOwnedByWrongProgram)
+                .with_pubkeys((*account_info.owner, TickArray::owner())),
+        );
+    }
+
+    let data = account_info.try_borrow_data()?;
+    if data.len() < TickArray::discriminator().len() {
+        return Err(anchor_lang::error::ErrorCode::AccountDiscriminatorNotFound.into());
+    }
+
+    let disc_bytes = arrayref::array_ref![data, 0, 8];
+    if disc_bytes != &TickArray::discriminator() {
+        return Err(anchor_lang::error::ErrorCode::AccountDiscriminatorMismatch.into());
+    }
+
+    let tick_array: Ref<TickArray> = Ref::map(data, |data| {
+        bytemuck::from_bytes(&data[8..std::mem::size_of::<TickArray>() + 8])
+    });
+
+    let start_tick_index = tick_array.start_tick_index;
+    let whirlpool = tick_array.whirlpool;
+    drop(tick_array);
+
+    Ok(TickArrayAccount::Initialized {
+        tick_array_whirlpool: whirlpool,
+        start_tick_index,
+        account_info,
+    })
+}
+
+fn get_start_tick_indexes(whirlpool: &Account<Whirlpool>, a_to_b: bool) -> Vec<i32> {
+    let tick_current_index = whirlpool.tick_current_index;
+    let tick_spacing = whirlpool.tick_spacing as i32;
+    let ticks_in_array = TICK_ARRAY_SIZE * tick_spacing;
+
+    let start_tick_index_base = floor_division(tick_current_index, ticks_in_array) * ticks_in_array;
+    let offset = if a_to_b {
+        [0, -1, -2]
+    } else {
+        let shifted = tick_current_index + tick_spacing >= start_tick_index_base + ticks_in_array;
+        if shifted {
+            [1, 2, 3]
+        } else {
+            [0, 1, 2]
+        }
+    };
+
+    let start_tick_indexes = offset
+        .iter()
+        .filter_map(|&o| {
+            let start_tick_index = start_tick_index_base + o * ticks_in_array;
+            if is_valid_start_tick_index(start_tick_index, ticks_in_array) {
+                Some(start_tick_index)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<i32>>();
+
+    start_tick_indexes
+}
+
+fn is_valid_start_tick_index(start_tick_index: i32, ticks_in_array: i32) -> bool {
+    // caller must assure that start_tick_index is multiple of ticks_in_array
+    start_tick_index + ticks_in_array > MIN_TICK_INDEX && start_tick_index < MAX_TICK_INDEX
```

Comment:
hmm, I think there is a discrepancy between this and the existing `check_is_valid_start_tick`, I believe the existing `check_is_valid_start_tick` uses an equality condition on the`MIN_TICK_INDEX`, i.e. `start_tick_index + ticks_in_array >= MIN_TICK_INDEX`?

Reply:
Good catch!
will check this edge case.

Good news is that 443636 (abs value of MIN_TICK_INDEX) is not multileple of 88.
So this MIN_TICK_INDEX will never be start_tick_index. 😁

Reply:
I chose the simplest approach.
Use `Tick::check_is_valid_start_tick` 😁